mono/profiler/proflog.c

```diff
@@ -4141,25 +4006,7 @@ helper_thread (void* arg)
 
 		if (FD_ISSET (prof->pipes [0], &rfds)) {
 			char c;
-			int r = read (prof->pipes [0], &c, 1);
-			if (r == 1 && c == 0) {
-				StatBuffer *sbufbase = prof->stat_buffers;
-				StatBuffer *sbuf;
-				if (!sbufbase->next)
-					continue;
-				sbuf = sbufbase->next->next;
-				sbufbase->next->next = NULL;
-				if (do_debug)
-					fprintf (stderr, "stat buffer dump\n");
-				if (sbuf) {
-					dump_sample_hits (prof, sbuf);
-					free_buffer (sbuf, sbuf->size);
-					safe_send_threadless (prof, ensure_logbuf (0));
-				}
-				continue;
-			}
-			/* time to shut down */
-			dump_sample_hits (prof, prof->stat_buffers);
+			read (prof->pipes [0], &c, 1);
```

Comment:
The helper thread does several other things before getting back to checking the samplehit pipe. In my version of this patch (the one we did not go with) this seemed to introduce problems once the frequency of hits to this pipe increases by a few hundred times (because we are no longer batching them in a samplebuf)…